### PR TITLE
Add improvements to the custom types section in doc

### PR DESCRIPTION
**Pull request by javmorcas**
This pull request is coming from the issue #327. When the pull request is merged into "master" the issue will be [resolved, closed, fixed] #XX

### CODE REVIEW

File: srcDoc/custom_types.rst [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+
+============
+Custom Types
+============
+
+*RTI Perftest* allows the user to add their own `idl` types, instead of using
+the one provided by default. The purpose of this feature is to get a closer
+approximation to the performance behavior of the user's final system. The main
+reason why using one type or another will result in different outputs for latency
+and throughput is the serialization and deserialization times.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:11]: *RTI Perftest* allows the user to test the performance of their own datatypes defined via 'idl', instead of 
the one provided by default. The purpose of this feature is to get a closer
approximation to the performance behavior of the user's final system. The main
reason why using one type or another will result in different results for latency
and throughput is that serialization and deserialization performance depends on the complexity of the datatype.

File: srcDoc/custom_types.rst [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+
+This feature is designed to work for regular data, *DynamicData*
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:13]: This feature is designed to work for regular data, *DynamicData*
and *FlatData* type definitions (required to test Connext DDS FlatData and/or Shared Memory Zero Copy features).

File: srcDoc/custom_types.rst [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+and *FlatData* types.
+
+**Note 1:** For *DynamicData*, the way this is implemented is by representing the data
+as a regular *idl* generated type, and then, internally use the *DynamicData*
+type support representation equivalent for that type as well as build the samples
+with the *DynamicData* API. Hence, providing the *idl* structure is required to the user.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:19]: *DynamicData* testing uses the Connext DDS *DynamicData* API to serialize/deserialize/copy and otherwise manipulate a data structure instead of compiled type support-generated functions from the user's type definition file.  However, for Perftest to do this with a user's datatype, the user must still provide their datatype definition in an *idl* file with which *DynamicData* will use a compiled/generated function to access the typecode for the datatype.

File: srcDoc/custom_types.rst [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+
+**Note 2:** The use of *Flat Data* requires the generation of a regular type as
+well due to the current implementation. See especific notes in a dedicated section
+below.
+
+To simplify testing using this feature. Some working examples can be found and
+used under `resource/exaples_customType`.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:26]: resource/examples_customType

File: srcDoc/custom_types.rst [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+and *FlatData* types.
+
+**Note 1:** For *DynamicData*, the way this is implemented is by representing the data
+as a regular *idl* generated type, and then, internally use the *DynamicData*
+type support representation equivalent for that type as well as build the samples
+with the *DynamicData* API. Hence, providing the *idl* structure is required to the user.
+
+**Note 2:** The use of *Flat Data* requires the generation of a regular type as
+well due to the current implementation. See especific notes in a dedicated section
+below.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:23]: **Note 2:** To test *Flat Data* serialization/deserialization of user data types requires the user to define both a regular datatype and a *FlatData* variant.  See specific notes in the section **Custom Types + `FlatData`** below.

File: srcDoc/custom_types.rst [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+
+To simplify testing using this feature. Some working examples can be found and
+used under `resource/exaples_customType`.
+
+How does it work
+~~~~~~~~~~~~~~~~
+
+*RTI Perftest* default type resembles the following (with some variations depending on
+the use of *Keyed data*, *Flat Data*, *Unbounded Types* or *Zero Copy*):
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:32]: *RTI Perftest* default datatype

File: srcDoc/custom_types.rst [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+        octet key[KEY_SIZE];
+        long entity_id;
+        unsigned long seq_num;
+        long timestamp_sec;
+        unsigned long timestamp_usec;
+        long latency_ping;
+        sequence<octet, MAX_BOUNDED_SEQ_SIZE> bin_data;
+    };
+
+However, in order to use a given type, the struct has been modified as follows:
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:46]: However, in order to use an user-specified datatype,

File: srcDoc/custom_types.rst [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+.. code-block:: c
+
+    struct TestData_t {
+        octet key[KEY_SIZE];
+        long entity_id;
+        unsigned long seq_num;
+        long timestamp_sec;
+        unsigned long timestamp_usec;
+        long latency_ping;
+        sequence<octet, MAX_BOUNDED_SEQ_SIZE> bin_data;
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:57]: Seems to me that can remove this member if custom data is being used, so something like
#ifndef RTI_CUSTOM_TYPE
        sequence<octet, MAX_BOUNDED_SEQ_SIZE> bin_data;
#endif

File: srcDoc/custom_types.rst [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+    #ifdef RTI_CUSTOM_TYPE
+        RTI_CUSTOM_TYPE custom_type;
+        // custom_type_size is the serialize size of the field custom_type
+        // in the the sample that we send on the wire
+        long custom_type_size;
+    #endif
+    };
+
+Notice that a `RTI_CUSTOM_TYPE` member has been added to the struct (conditionally).
+This placeholder is replaced at generation time with the type specified by the
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:67]: This macro is replaced

File: srcDoc/custom_types.rst [hunk 10]
```diff
@@ -0,0 +1,578 @@
...
+Therefore if the customer provides an `.idl` file with the following structure:
+
+.. code-block:: c
+
+    struct Test {
+        sequence<octet, MAX_SIZE> test_seq;
+        long test_long; //@key
+    };
+
+The type Name needed for `RTI_CUSTOM_TYPE` will be `Test` (independently
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:79]: Users must supply the name 'Test' to the perftest build process as the name of their customType so that 'RTI_CUSTOM_TYPE' will be replaced by 'Test'.

Not sure if "(independently on the name of the file where the struct is, or if it is preceeded or followed by
other structures)." is useful or (even understandable at this point in the doc)".  I would probably delete it.

File: srcDoc/custom_types.rst [hunk 11]
```diff
@@ -0,0 +1,578 @@
...
+on the name of the file where the struct is, or if it is preceeded or followed by
+other structures).
+
+Steps to use Custom Types
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The steps needed to use a custom type in *RTI Perftest* are:
+
+Copy the IDL files into `~/rtiperftest/srcIdl/customTypes/`
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:88]: Copy the IDL file(s) used to define the user's datatype to be tested into `~/rtiperftest/srcIdl/customTypes/`.

File: srcDoc/custom_types.rst [hunk 12]
```diff
@@ -0,0 +1,578 @@
...
+-----------------------------------------------------------
+
+There are some restrictions for this step:
+
+- You can include multiple `.idl` files, but all of them should be at the same
+  level (no folders are allowed inside the `customTypes` folder.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:94]: missing end ')'

File: srcDoc/custom_types.rst [hunk 13]
```diff
@@ -0,0 +1,578 @@
...
+- The `idl` files cannot be named `custom.idl`. A file with this name will be generated
+  automatically by the `build` script, so this file would be overwritten.
+- The main type you will link should be declared as ``struct <NameOfTheType> {``.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:97]: The customType that you want to test must be declared as in an '.idl' file as ``struct <NameOfTheType> {``.
In the declaration, there *must* be only a single space character between `struct` and `{` and the name of the datatype. The build scripts will complain that they can't find your datatype definition in the '.idl' files if your datatype definition does not follow the required format.


spelling/typo:  "accross" --> "across"

File: srcDoc/custom_types.rst (hunk 1 above)
Comment (howard-rti) [reply srcDoc/custom_types.rst:11]: By the way, is the testing of customDatatypes only supported in traditional C++?  Not in C, Modern C++ or Java?  If that is the case, should state that here as well.

File: srcDoc/custom_types.rst [hunk 14]
```diff
@@ -0,0 +1,578 @@
...
+You should find a file in `${PERFTEST_HOME}/srcCpp` named `customType.cxx`. This
+file contains several functions that must be implemented in order to correctly
+initialize, finalize and populate the sample.
+
+**Note:** *RTI Perftest* will not initialize by default sequences, optional members
+or non-primitive structures. This means that those fields will need to be
+initialized by the user in the `set_custom_type_data`, `set_custom_type_dynamic_data`
+and `set_custom_type_data_flatdata`.
+
+**Note:** If using *FlatData* see the dedicated section below.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:113]: **Note:** If using *FlatData* see the section **Custom Types + `FlatData`** below.

File: srcDoc/custom_types.rst [hunk 15]
```diff
@@ -0,0 +1,578 @@
...
+
+Run the `build` script adding ``--customType <NameOfTheType``
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:115]: missing ">" in "<NameofTypeType"

File: srcDoc/custom_types.rst [hunk 16]
```diff
@@ -0,0 +1,578 @@
...
+-------------------------------------------------------------
+
+Run the ``build.sh`` or ``build.bat`` script using the ``--customType`` option.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:118]: using the ``--customType <NameOfTheType>`` option.

File: srcDoc/custom_types.rst [hunk 17]
```diff
@@ -0,0 +1,578 @@
...
+- The main type you will link should be declared as ``struct <NameOfTheType> {``.
+  The `struct` and `{` words should be spaced as shows. This way the build scripts
+  will be able to find the type accross a list of `.idl`.
+
+Implement the API custom type functions of `customType.cxx`
+-----------------------------------------------------------
+
+You should find a file in `${PERFTEST_HOME}/srcCpp` named `customType.cxx`. This
+file contains several functions that must be implemented in order to correctly
+initialize, finalize and populate the sample.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:106]: Here, the doc is stating "must be implemented", later on it's stated 

"The following functions should be implemented (optionally) to properly
   initialize and set the custom type structures."

I would suggest

You should find a file in `${PERFTEST_HOME}/srcCpp/customType.cxx`. This
file contains several functions that may need to be implemented in order to correctly
initialize, finalize and populate the sample.

File: srcDoc/custom_types.rst [hunk 18]
```diff
@@ -0,0 +1,578 @@
...
+
+**Note:** *RTI Perftest* will not initialize by default sequences, optional members
+or non-primitive structures. This means that those fields will need to be
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:109]: I'm not sure what "non-primitive structures" are.  Do you mean "non-primitive" members?  If so, do you mean members that are pointers?  Does this include strings (aka "char *")?  Or do you actually mean a member that's a nested structure (but not a pointer to a nested structure)?

For _regular_ custom type testing, if the perftest code calls TestData_tTypeSupport::create_data(), wouldn't non-@optional pointers/strings, etc. of the nested user customType all be allocated?  And contained sequences initialized to hold max length elements with a current length of 0?

If so, I would probably modify this to state

**Note:** Users must initialize or otherwise set the values of certain types of members of their custom datatype or else those members may not be sent during testing.  Theses includes setting the length of sequence members (default length is 0), the values of strings (default is empty string ""), and the values of any @optional members that you want to be sent during tests.  To set those values, you may need to customize the implementations of `set/initialize_custom_type_data`, `set/intialize_custom_type_dynamic_data` and/or `set/initialize_custom_type_data_flatdata`.

File: srcDoc/custom_types.rst (hunk 16 above)
Comment (howard-rti) [reply srcDoc/custom_types.rst:118]: The title of this section is "Run the `build` script adding ``--customType <NameOfTheType``" and then the first line in this section is "Run the ``build.sh`` or ``build.bat`` script using the ``--customType`` option."  Seems a bit redundant.

Recommend

Building Perftest with custom types
-----------------------------------------------------------------------------------

Run the ``build.sh`` or ``build.bat`` script with the additional ``--customType <NameOfTheType>`` option.  For testing FlatData and/or ZeroCopy, please see the section **Custom Types + `FlatData`**.

File: srcDoc/custom_types.rst [hunk 19]
```diff
@@ -0,0 +1,578 @@
...
+and `set_custom_type_data_flatdata`.
+
+**Note:** If using *FlatData* see the dedicated section below.
+
+Run the `build` script adding ``--customType <NameOfTheType``
+-------------------------------------------------------------
+
+Run the ``build.sh`` or ``build.bat`` script using the ``--customType`` option.
+
+**Note:** If using *FlatData* see the dedicated section below.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:120]: Would not be needed if modification is made to the beginning of this section is adopted.

File: srcDoc/custom_types.rst [hunk 20]
```diff
@@ -0,0 +1,578 @@
...
+
+Run *RTI Perftest* as usual
+---------------------------
+
+A string at the beginning of the execution should indicate that you are using a
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:125]: To verify that _perftest_cpp_ is testing a customType, look for output similar to this when the test is started:

RTI Perftest Master df3055c (RTI Connext DDS 6.0.1)

**Custom Type provided: 'Test'**

Perftest Configuration:
        Reliability: Reliable
        Keyed: No
...

File: srcDoc/custom_types.rst [hunk 21]
```diff
@@ -0,0 +1,578 @@
...
+custom type and that this is not a regular execution.
+
+
+Custom Types + `FlatData`
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+*FlatData* types are at this point in *RTI Perftest* associated to regular types.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:132]: I don't think this sentence is needed.

*FlatData* types are at this point in *RTI Perftest* associated to regular types.

File: srcDoc/custom_types.rst [hunk 22]
```diff
@@ -0,0 +1,578 @@
...
+If you want to use *FlatData* with for your Custom Types, follow these
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:133]: If you want to test your customType using the *FlatData* language binding, follow these

File: srcDoc/custom_types.rst [hunk 23]
```diff
@@ -0,0 +1,578 @@
...
+additional steps:
+
+- Create a copy of your original type (which should not be `FlatData`) in the same IDL file.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:136]: Create an identical copy of your customType in the same IDL file and rename the datatype, for example from "Test" to "Test_flatdata".

File: srcDoc/custom_types.rst [hunk 24]
```diff
@@ -0,0 +1,578 @@
...
+- Make this new type mutable (``@mutable``) and make it FlatData compatible (``@language_binding(FLAT_DATA)``).
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:137]: `@mutable
@language_binding(FLAT_DATA)
struct Test_flatdata {
...
`

File: srcDoc/custom_types.rst [hunk 25]
```diff
@@ -0,0 +1,578 @@
...
+- Implement the additional API custom type functions of customtype.cxx for *FlatData* types.
+  This is a *must*, as the type is ``@mutable``, and every field is considered as optional
+  and not sent into the wire if it is not set.
+- Run the build script with the command-line parameter ``--customType <type>`` and ``--customTypeFlatData <flat_type>``.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:141]: parameter --> parameters

File: srcDoc/custom_types.rst [hunk 26]
```diff
@@ -0,0 +1,578 @@
...
+- Run *RTI Perftest* as usual.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:142]: Run *RTI Perfest* with ``-flatData`` to test using FlatData with your customType and/or ``-flatData -zeroCopy`` to test Zero Copy over shared memory with your customType.

File: srcDoc/custom_types.rst [hunk 27]
```diff
@@ -0,0 +1,578 @@
...
+
+Full example using Custom Types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following types will be used for this example. The code can be found under
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:147]: The code can be found under  -->  Related files can be found under

File: srcDoc/custom_types.rst [hunk 28]
```diff
@@ -0,0 +1,578 @@
...
+
+.. code-block:: c
+
+    const long SIZE_TEST_SEQ = 100;
+    const long SIZE_TEST_STRING = 128;
+
+    enum TestEnum {
+        ENUM1,
+        ENUM2
+    };//@Extensibility FINAL_EXTENSIBILITY
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:158]: I suggest modifying this RTI-proprietary method of specifying extensibility to the standard one. 

@final

File: srcDoc/custom_types.rst [hunk 29]
```diff
@@ -0,0 +1,578 @@
...
+
+    struct Test {
+        long test_long;
+        TestEnum test_enum;
+        StringTest test_string;
+        SeqTest test_seq;
+    };//@Extensibility FINAL_EXTENSIBILITY
+
+These functions in the example are spreaded across several `idl` files to show the
+multiple files capability of the feature.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:176]: For the example, these definitions are spread across several `idl` files to demonstrate the ability of perftest to use multiple files for defining a customType.

File: srcDoc/custom_types.rst [hunk 30]
```diff
@@ -0,0 +1,578 @@
...
+
+These are the steps needed to use the above type in *RTI Perftest* for the
+C++ (Traditional) API:
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:179]: By making this statement, it begs the question, how do I do it for other languages, C, C++ (Modern), Java.  If they are not supported, then this statement is better phrased:

The following steps are used to build and run *RTI Perftest* to test the customType `Test` defined above.  Note, the testing of customTypes is only supported using the DDS C++ (traditional) API.

File: srcDoc/custom_types.rst [hunk 31]
```diff
@@ -0,0 +1,578 @@
...
+  This is a *must*, as the type is ``@mutable``, and every field is considered as optional
+  and not sent into the wire if it is not set.
+- Run the build script with the command-line parameter ``--customType <type>`` and ``--customTypeFlatData <flat_type>``.
+- Run *RTI Perftest* as usual.
+
+Full example using Custom Types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following types will be used for this example. The code can be found under
+``resource/examples_customType/no_key_large``.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:148]: In my opinion, this isn't a great example to document...it's not really a "Full example".  It's not a keyed datatype nor does it show flat data.

File: srcDoc/custom_types.rst [hunk 32]
```diff
@@ -0,0 +1,578 @@
...
+
+These functions in the example are spreaded across several `idl` files to show the
+multiple files capability of the feature.
+
+These are the steps needed to use the above type in *RTI Perftest* for the
+C++ (Traditional) API:
+
+1. Copy all the `idl` files into the `srcIdl/customType/` folder.
+
+2. The following functions should be implemented (optionally) to properly
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:183]: The file ``resource/examples_customType/no_key_large/srcCpp/CustomType.cxx`` has been customized to implement some required functions needed to initialize members of the data structure.  For your own customType, you would modify the implementations of the functions in "rtiperftest/srcCpp/CustomType.cxx" as required.  To test the type defined in the example ``resource/examples_customType/no_key_large/``, you only need to overwrite ``rtiperftest/srcCpp/CustomType.cxx`` with ```resource/examples_customType/no_key_large/srcCpp/CustomType.cxx``.

File: srcDoc/custom_types.rst [hunk 33]
```diff
@@ -0,0 +1,578 @@
...
+   initialize and set the custom type structures.
+
+    - **initialize_custom_type_data**
+    This function is used to initialize your data.
+    Using this function, you will be able to allocate memory or set an immutable
+    field of the data.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:189]: This function is called once to initialize the instance of your datastructure sent during the test.
    Using this function, you will be able to allocate memory or set an immutable field of the data.

Use this method to set any @optional members that you want to be sent.  
Use this method to set a fix length of sequence members.

File: srcDoc/custom_types.rst [hunk 34]
```diff
@@ -0,0 +1,578 @@
...
+    The function takes one argument:
+
+        - A reference to custom type data.
+
+    .. code-block:: c
+
+        bool initialize_custom_type_data(RTI_CUSTOM_TYPE &data)
+        {
+            bool success = true;
+            if (!data.test_seq.test_seq.ensure_length(SIZE_TEST_SEQ, SIZE_TEST_SEQ)) {
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:199]: I don't quite understand why this example is setting a compiled-in, non runtime changeable length for the sequence.  It seems that the string member should be initialized here.  And the length should be set in the set_custom_type_data() method.

File: srcDoc/custom_types.rst (hunk 1 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:11]: Done.

File: srcDoc/custom_types.rst (hunk 2 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:13]: Done.

File: srcDoc/custom_types.rst (hunk 3 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:19]: Done.

File: srcDoc/custom_types.rst (hunk 4 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:26]: Done.

File: srcDoc/custom_types.rst (hunk 6 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:32]: Done.

File: srcDoc/custom_types.rst (hunk 7 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:46]: Done.

File: srcDoc/custom_types.rst (hunk 8 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:57]: I think this is something we can study. I will work on it in a different branch

File: srcDoc/custom_types.rst (hunk 9 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:67]: Done.

File: srcDoc/custom_types.rst (hunk 10 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:79]: Done.

File: srcDoc/custom_types.rst (hunk 11 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:88]: Done.

File: srcDoc/custom_types.rst (hunk 12 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:94]: Done.

File: srcDoc/custom_types.rst (hunk 13 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:97]: Done

File: srcDoc/custom_types.rst (hunk 17 above)
Comment (javmorcas) [reply srcDoc/custom_types.rst:106]: Done.

File: srcDoc/custom_types.rst [hunk 35]
```diff
@@ -0,0 +1,578 @@
...
+                success = false;
+            }
+            data.test_enum = ENUM1;
+            return success;
+        }
+
+    - **register_custom_type_data**
+    This function is used to set your data before being registered. It is only
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:225]: I'm not sure I understand the use and thus requirements for this function.  

So this is only needed for tests using keys, which would use the TestDataKeyed_t (and variants)?

So, for a keyed test, the test will create keys in the form of unsigned long values that this function will need to store into the key fields defined by the user data type if it is defined with a key field(s)?  Is that the idea?  But why is this needed?  Since the octet key member is still a key member, couldn't the test framework just continue to set that value itself without requiring setting of any key members inside the user structure and still create the number of instances required for the test?

In short, I don't think this function is required to support the testing of keys when using user data types...

File: srcDoc/custom_types.rst [hunk 36]
```diff
@@ -0,0 +1,578 @@
...
+
+    - **set_custom_type_data**
+    This function is used to set your data before it is sent.
+    It is called every time the data is sent.
+    You must set the custom type data before it is sent with the right
+    "key" value and the "targetDataLen".
+    The function takes three arguments:
+
+        - A reference to custom type data.
+        - A specific number unique for every key.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:249]: Again, I'm not sure that there's any additional functionality added with this key argument.  The key value could have just been set by the test framework in it's own "key" member and not rely on this method to set a key value in the customType.

File: srcDoc/custom_types.rst [hunk 37]
```diff
@@ -0,0 +1,578 @@
...
+        void register_custom_type_data(RTI_CUSTOM_TYPE & data, unsigned long key)
+        {
+            data.test_long = key;
+        }
+
+    - **set_custom_type_data**
+    This function is used to set your data before it is sent.
+    It is called every time the data is sent.
+    You must set the custom type data before it is sent with the right
+    "key" value and the "targetDataLen".
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:245]: You must implement this method if you want perftest to dynamically set the sizes of the variable length members of your data structure such as sequences or strings.

File: srcDoc/custom_types.rst [hunk 38]
```diff
@@ -0,0 +1,578 @@
...
+            bool success = true;
+            data.test_long = key;
+            if (sprintf(data.test_string.test_string, "Hello World! %lu", key) < 0) {
+                success = false;
+            }
+            return success;
+        }
+
+    - **finalize_custom_type_data**:
+    This function is used to remove your data. It is called in the destructor.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:270]: I'm actually unsure why this finalize is needed?  Only if you're running some sort of memory checker/validation software for memory leaks?

Yes, it is programming "correctness" to have such a function.  But practically, speaking, there's no reason to implement this.  Having it available is mostly just wasted effort (by us to support/document it) other than satisfying those folks who are OCD...

File: srcDoc/custom_types.rst [hunk 39]
```diff
@@ -0,0 +1,578 @@
...
+            if (!data.test_seq.test_seq.maximum(0)) {
+                success = false;
+            }
+            return success;
+        }
+
+    - **initialize_custom_type_dynamic_data**:
+    This function is used to initialize your DynamicData.
+    Using this function, you will be able to allocate memory or set an immutable
+    field of the data.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:289]: Same comments as for **initialize_custom_type_data**.

File: srcDoc/custom_types.rst [hunk 40]
```diff
@@ -0,0 +1,578 @@
...
+    .. code-block:: c
+
+        bool set_custom_type_data(
+                RTI_CUSTOM_TYPE & data,
+                unsigned long key,
+                int targetDataLen)
+        {
+            bool success = true;
+            data.test_long = key;
+            if (sprintf(data.test_string.test_string, "Hello World! %lu", key) < 0) {
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:263]: I would switch this with the contents of initialize.  The string should have been set once.  There's really no reason to set the test_string, over and over again.  It's not a key...well, I guess it would be in the key test since the "Test" struct itself has no keys, and the TestDataKeyed_t struct declares the nested "Test" struct as keyed, and thus all members are part of the key.  But that's not the point.  The point is that the embedded string probably should only need to be initialized once (sprintf is pretty expensive to call, and the test isn't really supposed to be measuring the cost of setting the value of the data being sent).  

On the other hand, this would be the logical place to use the targetDataLen to set the length of the sequence member.

File: srcDoc/custom_types.rst [hunk 41]
```diff
@@ -0,0 +1,578 @@
...
+            if (retcode != DDS_RETCODE_OK) {
+                fprintf(stderr, "set_long(test_long) failed: %d.\n", retcode);
+            }
+        }
+
+    - **set_custom_type_dynamic_data**:
+    This function is used to set your DynamicData before it is sent.
+    It is called every time the data is sent.
+    Set the custom type data before it is sent with the right "key"
+    value and the "targetDataLen".
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:333]: Same comments as in **set_custom_type_data**.

File: srcDoc/custom_types.rst [hunk 42]
```diff
@@ -0,0 +1,578 @@
...
+            }
+            retcode = testSeqData.set_long_seq(
+                        "test_seq",
+                        DDS_DYNAMIC_DATA_MEMBER_ID_UNSPECIFIED,
+                        longSeq);
+            if (retcode != DDS_RETCODE_OK) {
+                fprintf(stderr, "set_long(test_seq) failed: %d.\n", retcode);
+                success = false;
+            }
+            retcode = customTypeData.unbind_complex_member(testSeqData);
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:414]: Is the unbinding required to finalize the setting of the data?

If so, would be useful to have a comment making that statement.  Not sure how familiar people will be with the proper usage of the DynamicData API (I myself am a neophyte in this respect).

File: srcDoc/custom_types.rst [hunk 43]
```diff
@@ -0,0 +1,578 @@
...
+        Waiting to discover 1 publishers ...
+        Waiting for data ...
+        Length:   464  Packets: 68081040  Packets/s(ave): 1134692  Mbps(ave):  4212.0  Lost:     0 (0.00%)
+        Finishing test...
+        Test ended.
+
+
+
+5. You can also launch *RTI Perftest* with your customType using DynamicData.
+
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:530]: In this mode, Perftest will use the DynamicData API to serialize/deserialize the customType data structure instead of the generated serialization/deserialization code.

File: srcDoc/custom_types.rst [hunk 44]
```diff
@@ -0,0 +1,578 @@
...
+        Transport Configuration:
+            Kind: UDPv4 & SHMEM (taken from QoS XML file)
+            Use Multicast: False
+
+        Waiting to discover 1 publishers ...
+        Waiting for data ...
+        Length:   464  Packets:  8146078  Packets/s(ave):  135770  Mbps(ave):   504.0  Lost:     0 (0.00%)
+        Finishing test...
+        Test ended.
```
Comment (howard-rti) [inline srcDoc/custom_types.rst:596]: So, I would probably omit the DynamicData example and use a FlatData/ZeroCopy example instead.  I've not heard that customers are interested in the performance of DynamicData on their datatypes....not even sure what their use case would be myself.  But certainly, I have seen interest in FlatData (and mainly ZeroCopy which requires FlatData).
